github.com/cockroachdb/pebble/internal/metamorphic: TestMetaTwoInstance failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/16645634?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/16645634?buildTab=artifacts#meta) on refs/heads/master @ [34e929e1fcac](https://github.com/cockroachdb/pebble/commits/34e929e1fcac):


```
        iter37.Next("") // [false] <nil> #3850
        iter36.SeekPrefixGE("yvodk@7") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #3851
        iter38 = iter37.Clone(false, "hmwupnwtnwoj", "jlnvguh@11", 1 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */) // <nil> #3852
        iter38.Next("") // [false] <nil> #3853
        db1.Get("ltaktczhi@9") // [""] pebble: not found #3854
        iter36.SeekPrefixGE("avncedojxmw@4") // [true,"avncedojxmw@4",<no point>,["avncedojxmw","avncedojxmw\x00")=>{"@4"="gmdsvaiuetjto"}*] <nil> #3855
        iter36.SeekGE("nggfnk", "") // [false] <nil> #3856
        iter37.Prev("") // [false] <nil> #3857
        iter37.Next("") // [false] <nil> #3858
        iter38.SeekPrefixGE("mrfd") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #3859
        db2.Set("nksptayrqmoi@11", "ehpuiwcrixfdq") // <nil> #3860
        iter38.First() // [false] <nil> #3861
        iter39 = iter36.Clone(false, "ahwuqfle@4", "dafszn", 2 /* key types */, 0, 0, false /* use L6 filters */, "@4" /* masking suffix */) // <nil> #3862
        iter39.SeekLT("lfyo", "") // [true,"cpql",<no point>,["cpql","dafszn")=>{"@4"="gmdsvaiuetjto"}*] <nil> #3863
        iter36.SeekPrefixGE("mdshz@2") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #3864
        iter39.SetBounds("ahwuqfle@4", "ahwuqfle@4") // <nil> #3865
        iter39.SeekLT("ahwuqfle@4", "") // [false] <nil> #3866
        iter39.Next("") // [false] <nil> #3867
        iter39.Prev("") // [false] <nil> #3868
        iter38.SeekGE("jzsyxhyha@9", "") // [false] <nil> #3869
        db2.Set("snkeitowwfpx", "ebbhek") // <nil> #3870
        iter36.SetBounds("ahwuqfle@4", "ahwuqfle@4") // <nil> #3871
        iter36.SeekLT("ahwuqfle@4", "") // [false] <nil> #3872
        iter36.Next("") // [false] <nil> #3873
        iter36.Prev("") // [false] <nil> #3874
        iter39.NextPrefix() // [false] NextPrefix not permitted with upper bound ahwuqfle@4 #3875
        iter37.Last() // [false] <nil> #3876
        db2.Get("rizbl@2") // [""] pebble: not found #3877
        iter40 = db1.NewIter("", "", 0 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */) // <nil> #3878
        iter38.First() // [false] <nil> #3879
        db2.Get("coynhjqxt@4") // [""] pebble: not found #3880
        iter39.Last() // [false] <nil> #3881
        db1.DeleteRange("dyjwsimbo@11", "jktikyrig@11") // <nil> #3882
        iter37.Last() // [false] <nil> #3883
        iter40.Next("") // [false] <nil> #3884
        iter38.NextPrefix() // [false] NextPrefix not permitted with upper bound jlnvguh@11 #3885
        iter38.SeekPrefixGE("atajgsuke@10") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #3886
        iter39.Prev("") // [false] <nil> #3887
        iter39.SeekLT("rhalnitnuprh@8", "") // [false] <nil> #3888
        iter39.SeekPrefixGE("lkylmmg@9") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #3889
        iter39.Prev("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #3890
        iter37.Next("") // [false] <nil> #3891
        iter39.SeekPrefixGE("uggkqhokp@3") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #3892
        // 04:36:22.208 INFO: [JOB 79] ingesting: sstable created 000071
        // 04:36:22.208 INFO: [JOB 80] WAL created 000072
        // 04:36:22.208 INFO: [JOB 81] flushing 1 memtable (4.3KB) to L0
        // 04:36:22.208 INFO: [JOB 81] flushing: sstable created 000073
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240828-043620.1301987344096/random-003 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-003 (0.41s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1724819780130833882 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: PEBBLE-252

--- a/vfs/mem_fs.go
+++ b/vfs/mem_fs.go
@@ -8,7 +8,6 @@ import (
 	"bytes"
 	"fmt"
 	"io"
-	"maps"
 	"math/rand"
 	"os"
 	"path"
@@ -138,7 +137,7 @@ type CrashCloneCfg struct {
 // latter is controlled by CrashCloneCfg.
 func (y *MemFS) CrashClone(cfg CrashCloneCfg) *MemFS {
 	if !y.crashable {
-		panic("not a strict MemFS")
+		panic("not a crashable MemFS")
 	}
 	// Block all modification operations while we clone.
 	y.cloneMu.Lock()
@@ -635,10 +634,11 @@ func newRootMemNode() *memNode {
 }
 
 func cloneChildren(f map[string]*memNode) map[string]*memNode {
-	if len(f) == 0 {
-		return make(map[string]*memNode)
+	m := make(map[string]*memNode)
+	for k, v := range f {
+		m[k] = v
 	}
-	return maps.Clone(f)
+	return m
 }
 
 func (f *memNode) dump(w *bytes.Buffer, level int, name string) {
@@ -671,6 +671,8 @@ func (f *memNode) dump(w *bytes.Buffer, level int, name string) {
 	}
 }
 
+// CrashClone creates a crash-consistent clone of the subtree rooted at f, and
+// returns the new subtree. cloneMu must be held (in write mode).
 func (f *memNode) CrashClone(cfg *CrashCloneCfg) *memNode {
 	newNode := &memNode{isDir: f.isDir}
 	if f.isDir {

